File: sdks/python/src/ratify_protocol/constraints.py
"""Constraint evaluation — mirrors Go's constraints.go exactly.

Every semantic must produce the same verdict for the same inputs, or
cross-language conformance fails.
"""
from __future__ import annotations

import math
from datetime import datetime, timezone
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

from .types import Constraint, DelegationCert, VerifierContext


def evaluate_constraints(
    cert: DelegationCert,
    ctx: VerifierContext | None,
    now_sec: int,
) -> str | None:
    """Run every Constraint on cert against the caller-supplied
    VerifierContext. Return None iff all pass; error message otherwise.

    Fail-closed: unknown Type or missing required context field causes
    rejection.
    """
    if ctx is None:
        ctx = VerifierContext()
    for i, c in enumerate(cert.constraints or []):
        err = _evaluate_constraint(c, cert.cert_id, ctx, now_sec)
        if err is not None:
            return f"constraint[{i}] ({c.type}): {err}"
    return None


def _evaluate_constraint(
    c: Constraint,
    cert_id: str,
    ctx: VerifierContext,
    now_sec: int,
) -> str | None:
    t = c.type
    if t == "geo_circle":
        if ctx.current_lat is None or ctx.current_lon is None:
            return "constraint_unverifiable: no current location in context"
        d = _haversine_meters(ctx.current_lat, ctx.current_lon, c.lat, c.lon)
        if d > c.radius_m:
            return f"outside allowed radius: {d:.1f}m > {c.radius_m:.1f}m"
        return None

    if t == "geo_polygon":
        if ctx.current_lat is None or ctx.current_lon is None:
            return "constraint_unverifiable: no current location in context"
        if not c.points or len(c.points) < 3:
            return "polygon has fewer than 3 points"
        # v1 polygon is defined over equirectangular projection — correct
        # for small regions, incorrect for anti-meridian-crossing shapes.
        # Fail closed rather than silently return wrong answers. SPEC §5.7.2
        # documents the v1 small-region limitation; geodesic semantics are v2.
        if _polygon_spans_antimeridian(c.points):
            return "geo_polygon spans >180° longitude — v1 semantics are undefined for anti-meridian-crossing polygons (see SPEC §5.7.2)"
        if not _point_in_polygon(ctx.current_lat, ctx.current_lon, c.points):
            return "outside allowed polygon"
        return None

    if t == "geo_bbox":
        if ctx.current_lat is None or ctx.current_lon is None:
            return "constraint_unverifiable: no current location in context"
        # Format mirrors Go's %.6f — cross-SDK error_reason parity requires it.
        if ctx.current_lat < c.min_lat or ctx.current_lat > c.max_lat:
            return f"latitude {ctx.current_lat:.6f} outside [{c.min_lat:.6f}, {c.max_lat:.6f}]"
        # Anti-meridian-aware longitude check (SPEC §5.7.2).
        #   min_lon <= max_lon → ordinary bbox, lon must be in [min_lon, max_lon]
        #   min_lon >  max_lon → bbox wraps the 180° meridian (e.g.,
        #     min_lon=170, max_lon=-170 = "from 170°E through 180 to -170°W")
        #     — inside iff lon >= min_lon OR lon <= max_lon.
        if c.min_lon <= c.max_lon:
            if ctx.current_lon < c.min_lon or ctx.current_lon > c.max_lon:
                return f"longitude {ctx.current_lon:.6f} outside [{c.min_lon:.6f}, {c.max_lon:.6f}]"
        else:
            if ctx.current_lon < c.min_lon and ctx.current_lon > c.max_lon:
                return f"longitude {ctx.current_lon:.6f} outside wrapped [{c.min_lon:.6f}, {c.max_lon:.6f}]"
        has_alt = c.min_alt_m != 0.0 or c.max_alt_m != 0.0
        if has_alt:
            if ctx.current_alt_m is None:
                return "constraint_unverifiable: no altitude in context but bbox has altitude bounds"
            if ctx.current_alt_m < c.min_alt_m or ctx.current_alt_m > c.max_alt_m:
                return f"altitude {ctx.current_alt_m:.1f}m outside [{c.min_alt_m:.1f}m, {c.max_alt_m:.1f}m]"
        return None

    if t == "time_window":
        if not c.tz or not c.start or not c.end:
            return "malformed time_window: tz/start/end required"
        start = _parse_hhmm(c.start)
        end = _parse_hhmm(c.end)
        if start is None:
            return f"bad start time: {c.start}"
        if end is None:
            return f"bad end time: {c.end}"
        try:
            zone = ZoneInfo(c.tz)
        except ZoneInfoNotFoundError:
            return f'unknown timezone "{c.tz}"'
        local = datetime.fromtimestamp(now_sec, tz=timezone.utc).astimezone(zone)
        cur = local.hour * 60 + local.minute
        if start <= end:
            if cur < start or cur > end:
                return f"current {local.hour:02d}:{local.minute:02d} outside [{c.start}, {c.end}] {c.tz}"
        else:
            # Wrapping window (e.g. 22:00 to 06:00).
            if cur < start and cur > end:
                return f"current {local.hour:02d}:{local.minute:02d} outside wrapped [{c.start}, {c.end}] {c.tz}"
        return None

    if t == "max_speed_mps":
        if ctx.current_speed_mps is None:
            return "constraint_unverifiable: no current speed in context"
        if ctx.current_speed_mps > c.max_mps:
            return f"speed {ctx.current_speed_mps:.2f}mps exceeds max {c.max_mps:.2f}mps"
        return None

    if t == "max_amount":
        if ctx.requested_amount is None or ctx.requested_currency is None:
            return "constraint_unverifiable: no requested amount in context"
        if ctx.requested_currency != c.currency:
            return f'currency mismatch: requested "{ctx.requested_currency}", constraint "{c.currency}"'
        if ctx.requested_amount > c.max_amount:
            return f"amount {ctx.requested_amount:.2f} {ctx.requested_currency} exceeds max {c.max_amount:.2f} {c.currency}"
        return None

    if t == "max_rate":
        if ctx.invocations_in_window is None:
            return "constraint_unverifiable: no rate counter in context"
        if c.count <= 0 or c.window_s <= 0:
            return "malformed max_rate: count and window_s must be positive"
        got = ctx.invocations_in_window(cert_id, c.window_s)
        if got >= c.count:
            return f"rate limit exceeded: {got} invocations in last {c.window_s}s (max {c.count})"
        return None

    # Sentinel prefix lets verify.py route to identity_status=constraint_unknown.
    # Matches Go reference unknownConstraintError shape.
    return f'constraint_unknown: unknown constraint type "{t}"'


# ---- helpers ----


def _haversine_meters(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Great-circle distance on a sphere (WGS-84 mean radius)."""
    earth_radius_m = 6371000.0
    rad = math.pi / 180
    d_lat = (lat2 - lat1) * rad
    d_lon = (lon2 - lon1) * rad
    a = (
        math.sin(d_lat / 2) ** 2
        + math.cos(lat1 * rad) * math.cos(lat2 * rad) * math.sin(d_lon / 2) ** 2
    )
    return 2 * earth_radius_m * math.asin(min(1.0, math.sqrt(a)))


def _polygon_spans_antimeridian(points: list[list[float]]) -> bool:
    """Return True if the polygon's longitudes span more than 180°.

    The only way to span more than half the globe in longitude is to
    cross the anti-meridian, and equirectangular ray-casting can't handle
    that — so we refuse these polygons up-front (fail-closed) rather than
    silently compute wrong inclusion. Mirrors Go's polygonSpansAntimeridian.
    """
    if len(points) < 3:
        return False
    min_lon = points[0][1]
    max_lon = points[0][1]
    for p in points:
        if p[1] < min_lon:
            min_lon = p[1]
        if p[1] > max_lon:
            max_lon = p[1]
    return (max_lon - min_lon) > 180


def _point_in_polygon(lat: float, lon: float, poly: list[list[float]]) -> bool:
    """Ray casting in equirectangular projection. Fine for small polygons."""
    inside = False
    n = len(poly)
    j = n - 1
    for i in range(n):
        yi, xi = poly[i][0], poly[i][1]  # lat, lon
        yj, xj = poly[j][0], poly[j][1]
        if ((yi > lat) != (yj > lat)) and lon < (xj - xi) * (lat - yi) / (yj - yi) + xi:
            inside = not inside
        j = i
    return inside


def _parse_hhmm(s: str) -> int | None:
    if len(s) != 5 or s[2] != ":":
        return None
    try:
        h = int(s[:2])
        m = int(s[3:])
    except ValueError:
        return None
    if h < 0 or h > 23 or m < 0 or m > 59:
        return None
    return h * 60 + m

File: constraints.go
package ratify

import (
	"fmt"
	"math"
	"time"
)

// evaluateConstraints runs every Constraint on cert against the caller-supplied
// VerifierContext. Returns nil iff all pass; a descriptive error otherwise.
// Fail-closed: an unknown Type or a constraint whose required context field
// is absent causes rejection.
func evaluateConstraints(cert *DelegationCert, ctx VerifierContext, now time.Time) error {
	for i, c := range cert.Constraints {
		if err := evaluateConstraint(c, cert.CertID, ctx, now); err != nil {
			return fmt.Errorf("constraint[%d] (%s): %w", i, c.Type, err)
		}
	}
	return nil
}

func evaluateConstraint(c Constraint, certID string, ctx VerifierContext, now time.Time) error {
	switch c.Type {
	case ConstraintGeoCircle:
		if !ctx.HasLocation {
			return errConstraintUnverifiable("no current location in context")
		}
		d := haversineMeters(ctx.CurrentLat, ctx.CurrentLon, c.Lat, c.Lon)
		if d > c.RadiusM {
			return fmt.Errorf("outside allowed radius: %.1fm > %.1fm", d, c.RadiusM)
		}
		return nil

	case ConstraintGeoPolygon:
		if !ctx.HasLocation {
			return errConstraintUnverifiable("no current location in context")
		}
		if len(c.Points) < 3 {
			return fmt.Errorf("polygon has fewer than 3 points")
		}
		// v1 polygon is defined over equirectangular projection — correct
		// for small regions, incorrect for anti-meridian-crossing shapes.
		// Fail closed rather than silently return wrong answers. SPEC §5.7.2
		// documents the v1 small-region limitation; geodesic semantics are v2.
		if polygonSpansAntimeridian(c.Points) {
			return fmt.Errorf("geo_polygon spans >180° longitude — v1 semantics are undefined for anti-meridian-crossing polygons (see SPEC §5.7.2)")
		}
		if !pointInPolygon(ctx.CurrentLat, ctx.CurrentLon, c.Points) {
			return fmt.Errorf("outside allowed polygon")
		}
		return nil

	case ConstraintGeoBBox:
		if !ctx.HasLocation {
			return errConstraintUnverifiable("no current location in context")
		}
		if ctx.CurrentLat < c.MinLat || ctx.CurrentLat > c.MaxLat {
			return fmt.Errorf("latitude %.6f outside [%.6f, %.6f]", ctx.CurrentLat, c.MinLat, c.MaxLat)
		}
		// Anti-meridian-aware longitude check (SPEC §5.7.2).
		//   MinLon <= MaxLon → ordinary bbox, lon must be in [MinLon, MaxLon]
		//   MinLon >  MaxLon → bbox wraps the 180° meridian (e.g.,
		//     MinLon=170, MaxLon=-170 = "from 170°E through 180 to -170°W")
		//     — inside iff lon >= MinLon OR lon <= MaxLon.
		if c.MinLon <= c.MaxLon {
			if ctx.CurrentLon < c.MinLon || ctx.CurrentLon > c.MaxLon {
				return fmt.Errorf("longitude %.6f outside [%.6f, %.6f]",
					ctx.CurrentLon, c.MinLon, c.MaxLon)
			}
		} else {
			if ctx.CurrentLon < c.MinLon && ctx.CurrentLon > c.MaxLon {
				return fmt.Errorf("longitude %.6f outside wrapped [%.6f, %.6f]",
					ctx.CurrentLon, c.MinLon, c.MaxLon)
			}
		}
		// Altitude bounds are ignored when both are zero.
		if c.MinAltM != 0 || c.MaxAltM != 0 {
			if ctx.CurrentAltM < c.MinAltM || ctx.CurrentAltM > c.MaxAltM {
				return fmt.Errorf("altitude %.1fm outside [%.1fm, %.1fm]", ctx.CurrentAltM, c.MinAltM, c.MaxAltM)
			}
		}
		return nil

	case ConstraintTimeWindow:
		if c.TZ == "" || c.Start == "" || c.End == "" {
			return fmt.Errorf("malformed time_window: tz/start/end required")
		}
		loc, err := time.LoadLocation(c.TZ)
		if err != nil {
			return fmt.Errorf("unknown timezone %q", c.TZ)
		}
		start, err := parseHHMM(c.Start)
		if err != nil {
			return fmt.Errorf("bad start time: %w", err)
		}
		end, err := parseHHMM(c.End)
		if err != nil {
			return fmt.Errorf("bad end time: %w", err)
		}
		local := now.In(loc)
		cur := local.Hour()*60 + local.Minute()
		if start <= end {
			// Non-wrapping window.
			if cur < start || cur > end {
				return fmt.Errorf("current %02d:%02d outside [%s, %s] %s",
					local.Hour(), local.Minute(), c.Start, c.End, c.TZ)
			}
		} else {
			// Wrapping window (e.g. 22:00 to 06:00).
			if cur < start && cur > end {
				return fmt.Errorf("current %02d:%02d outside wrapped [%s, %s] %s",
					local.Hour(), local.Minute(), c.Start, c.End, c.TZ)
			}
		}
		return nil

	case ConstraintMaxSpeedMps:
		if !ctx.HasSpeed {
			return errConstraintUnverifiable("no current speed in context")
		}
		if ctx.CurrentSpeedMps > c.MaxMps {
			return fmt.Errorf("speed %.2fmps exceeds max %.2fmps", ctx.CurrentSpeedMps, c.MaxMps)
		}
		return nil

	case ConstraintMaxAmount:
		if !ctx.HasAmount {
			return errConstraintUnverifiable("no requested amount in context")
		}
		if ctx.RequestedCurrency != c.Currency {
			return fmt.Errorf("currency mismatch: requested %q, constraint %q", ctx.RequestedCurrency, c.Currency)
		}
		if ctx.RequestedAmount > c.MaxAmount {
			return fmt.Errorf("amount %.2f %s exceeds max %.2f %s",
				ctx.RequestedAmount, ctx.RequestedCurrency, c.MaxAmount, c.Currency)
		}
		return nil

	case ConstraintMaxRate:
		if ctx.InvocationsInWindow == nil {
			return errConstraintUnverifiable("no rate counter in context")
		}
		if c.Count <= 0 || c.WindowS <= 0 {
			return fmt.Errorf("malformed max_rate: count and window_s must be positive")
		}
		got := ctx.InvocationsInWindow(certID, c.WindowS)
		if got >= c.Count {
			return fmt.Errorf("rate limit exceeded: %d invocations in last %ds (max %d)",
				got, c.WindowS, c.Count)
		}
		return nil

	default:
		return unknownConstraintError{kind: string(c.Type)}
	}
}

// ---- helpers ----

type unverifiableError struct{ reason string }

func (e unverifiableError) Error() string { return "constraint_unverifiable: " + e.reason }

func errConstraintUnverifiable(reason string) error { return unverifiableError{reason: reason} }

// unknownConstraintError is the sentinel for unknown constraint types. The
// verifier routes to IdentityStatusConstraintUnknown on it rather than
// collapsing to IdentityStatusConstraintDenied, so a cert that carries a
// constraint shape a given verifier version does not understand fails
// explicitly (fail-closed, visible in audit) instead of silently.
type unknownConstraintError struct{ kind string }

func (e unknownConstraintError) Error() string {
	return "constraint_unknown: unknown constraint type " + fmtQuoted(e.kind)
}

// fmtQuoted wraps a string in double quotes for stable error text without
// pulling in strconv's escaping surprises — the constraint-type string is
// always safe ASCII by v1 spec.
func fmtQuoted(s string) string { return "\"" + s + "\"" }

// haversineMeters — great-circle distance on a sphere (WGS-84 mean radius).
// Accurate enough for bounded-radius geofence checks at the meter scale.
func haversineMeters(lat1, lon1, lat2, lon2 float64) float64 {
	const earthRadiusM = 6371000.0
	rad := math.Pi / 180
	dLat := (lat2 - lat1) * rad
	dLon := (lon2 - lon1) * rad
	a := math.Sin(dLat/2)*math.Sin(dLat/2) +
		math.Cos(lat1*rad)*math.Cos(lat2*rad)*math.Sin(dLon/2)*math.Sin(dLon/2)
	return 2 * earthRadiusM * math.Asin(math.Min(1.0, math.Sqrt(a)))
}

// polygonSpansAntimeridian returns true if the polygon's longitudes span
// more than 180°. The only way to span more than half the globe in
// longitude is to cross the anti-meridian, and the equirectangular
// ray-casting below can't handle that — so we refuse these polygons
// up-front (fail-closed) rather than silently compute wrong inclusion.
func polygonSpansAntimeridian(points [][2]float64) bool {
	if len(points) < 3 {
		return false
	}
	minLon, maxLon := points[0][1], points[0][1]
	for _, p := range points {
		if p[1] < minLon {
			minLon = p[1]
		}
		if p[1] > maxLon {
			maxLon = p[1]
		}
	}
	return (maxLon - minLon) > 180
}

// pointInPolygon — standard ray-casting algorithm in equirectangular
// projection. Fine for small polygons (<~100 km diameter); for very large
// polygons use geodesic inclusion instead.
func pointInPolygon(lat, lon float64, poly [][2]float64) bool {
	inside := false
	n := len(poly)
	for i, j := 0, n-1; i < n; i++ {
		xi, yi := poly[i][1], poly[i][0] // lon, lat
		xj, yj := poly[j][1], poly[j][0]
		if ((yi > lat) != (yj > lat)) &&
			(lon < (xj-xi)*(lat-yi)/(yj-yi)+xi) {
			inside = !inside
		}
		j = i
	}
	return inside
}

// parseHHMM — "HH:MM" 24-hour clock → minutes-since-midnight.
func parseHHMM(s string) (int, error) {
	if len(s) != 5 || s[2] != ':' {
		return 0, fmt.Errorf("expected HH:MM, got %q", s)
	}
	h := int(s[0]-'0')*10 + int(s[1]-'0')
	m := int(s[3]-'0')*10 + int(s[4]-'0')
	if h < 0 || h > 23 || m < 0 || m > 59 {
		return 0, fmt.Errorf("out-of-range HH:MM %q", s)
	}
	return h*60 + m, nil
}
